triton-vm/src/table/constraint_circuit.rs

```diff
@@ -1060,6 +875,182 @@ impl<II: InputIndicator> ConstraintCircuitMonad<II> {
+                true => base_constraints.push(new_constraint),
+                false => ext_constraints.push(new_constraint),
+            }
+
+            // Treat roots of the multicircuit explicitly.
+            for circuit in multicircuit.iter_mut() {
+                if circuit.circuit.borrow().id == chosen_node_id {
+                    circuit.circuit = new_circuit.clone();
+                }
+            }
+        }
+
+        (base_constraints, ext_constraints)
+    }
+
+    /// Heuristically pick a node from the given multicircuit that is to be substituted with a new
+    /// variable. The ID of the chosen node is returned.
+    fn pick_node_to_substitute(
+        multicircuit: &[ConstraintCircuitMonad<II>],
+        target_degree: Degree,
+    ) -> usize {
+        if multicircuit.is_empty() {
+            panic!("Multicircuit must be non-empty in order to pick a node from it.");
+        }
+
+        let multicircuit = multicircuit
+            .iter()
+            .map(|c| c.clone().consume())
+            .collect_vec();
+        let all_nodes = Self::all_nodes_in_multicircuit(&multicircuit);
+        let all_nodes: HashSet<_> = HashSet::from_iter(all_nodes.into_iter());
+
+        // Only nodes with degree > target_degree need changing.
+        let high_degree_nodes = all_nodes
+            .into_iter()
+            .filter(|node| node.degree() > target_degree)
+            .collect_vec();
+
+        // Collect all the nodes where some substitution is necessary.
+        // Substituting a node of degree 1 is both pointless and can lead to infinite iteration.
+        let low_degree_nodes = Self::all_nodes_in_multicircuit(&high_degree_nodes)
+            .into_iter()
+            .filter(|node| node.degree() <= target_degree)
+            .filter(|node| node.degree() > 1)
+            .collect_vec();
+
+        // If the resulting list is empty, there is no way forward. Stop – panic time!
+        assert!(
+            !low_degree_nodes.is_empty(),
+            "Could not lower degree of circuit to target degree. This is a bug."
+        );
+
+        // Of the remaining nodes, keep the ones occurring the most often.
+        let mut nodes_and_occurrences = HashMap::new();
+        for node in low_degree_nodes.iter() {
+            *nodes_and_occurrences.entry(node).or_insert(0) += 1;
+        }
+        let max_occurrences = nodes_and_occurrences
+            .iter()
+            .map(|(_, &count)| count)
+            .max()
+            .unwrap();
+        nodes_and_occurrences.retain(|_, &mut count| count == max_occurrences);
+        let mut candidate_nodes = nodes_and_occurrences.keys().cloned().collect_vec();
+
+        let max_degree = candidate_nodes
+            .iter()
+            .map(|node| node.degree())
+            .max()
+            .unwrap();
+        candidate_nodes.retain(|node| node.degree() == max_degree);
+
+        // If there are still multiple nodes, pick any.
+        candidate_nodes[0].id
```

Comment:
Using a high degree node for the substitution would be a bug, because the constraint that is to be added to the AIR is then also of high degree. In my current code, newly constructed constraints are not intended for further reduction. Using a random low degree node works sometimes, but can lead to issues regarding the order in which the new variables are filled in.